### PR TITLE
Single peptide Functional analysis

### PR DESCRIPTION
**Pull request by beardhatcode**
Adds a table of matched GO-numbers and EC numbers to the single peptide page

- [x] Quick GO images
- [x] GO number tables (per field)
- [x] EC number tables 
- [x] EC treeview
- [x] Tooltips in protein matches
- [x] Clean and optimize code

Cherry-pick of 327edfe39dd65af6715f360ad30bec492e172c66

_[Original pull request](https://github.ugent.be/unipept/unipept/pull/625) by @beardhatcode on Fri Nov 17 2017 at 15:02._
_Merged by @beardhatcode on Mon May 28 2018 at 14:21._

### CODE REVIEW

**Comment by bmesuere** [comment]
Without looking at the code, some comments:
- [x] Split the functions tab into 2 tabs: GO and EC
- [x] Split the GO stuff into 3 sections for the namespaces
- [x] Per GO namespace, include a table/list of occurring GO terms with both the term and full name, sorted by occurrence
- [ ] Does the quickGO api allow to specify a color to highlight the nodes? i.e. can we include a quickgo image that is heatmap coded?
- [x] Tweak the message "Not enough EC code annotaions" to "Sorry, none of the matched proteins had EC annotations"
- [x] Would it make sense to do something similar for the ec numbers as for the go terms, but using the 6 high level categories?
- [x] Figure out if we can include the ec names (if we go for the high level categories: in the table, tooltips otherwise)
- [x] Does it make sense to include a treeview graph of ec numbers?

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71667) by @bmesuere on Fri Nov 17 2017 at 15:28._

**Comment by bmesuere** [comment]
image
image


_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71733) by @bmesuere on Mon Nov 20 2017 at 09:56._

**Comment by bmesuere** [comment]
Conclusions after this iteration:
- [x] GO: @beardhatcode: can we include the quickgo graphs per namespace? We must find a way to fit it into the UI, maybe we can use thumbnails that open the full image on click. 1 image per sections, or the 3 images at the top next to each other.
- [x] GO: we must use fixed column widths so the columns align for the 3 tables
- [x] GO: only show the top _x_ rows per namespace and add a button to show all?
- [x] EC: I don't think that splitting them into the 6 separate classes makes sense here, so we should put them into 1 big table, possibly only showing the top _x_ rows as we plan to do with the go terms
- [x] EC: add a treeview of the ec terms at the top of that tab
- [x] EC&GO: @beardhatcode: add those little bar chart background things from the multi search
- [ ] General: the protein matches tab should probably be the first tab
- [ ] General: should we merge the lineage tree and table tabs? maybe first check the result of combining treeview + table on the ec tab
- [ ] General: the description at the top should be rewritten to include ec and go information. I think the "common lineage" can be dropped from the description

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71734) by @bmesuere on Mon Nov 20 2017 at 09:58._

**Comment by beardhatcode** [comment]
- [x] Add QuickGo charts per GO domain. Thumbnails that expand seem a good idea to me. I'll try to get that to work today. 
- [x] Fixed column widths  
- [x] Use the `AmountTable` class to show the table

- I don't really know how you look for EC  numbers but from what I read on Wikipedia it feels like splitting it up in 6 categories is not useful. For the GO numbers splitting up is intuitive as someone could be looking for the Biological processes of a sample only (I think).

---
## UI suggestion
For the UI of the GO term I think something roughly like this may be better:
An overview of all GO terms, with tooltips that give more information (name)
screen shot 2017-11-20 at 11 01 29
And if you click `show more` than you would be shown the full table (only of the selected field)
screen shot 2017-11-20 at 11 01 38



_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71735) by @beardhatcode on Mon Nov 20 2017 at 11:10._

**Comment by bmesuere** [comment]
The GO numbers say nothing at all, so we should definitely always include the name, even in a summary.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71736) by @bmesuere on Mon Nov 20 2017 at 11:23._

**Comment by beardhatcode** [comment]
Fair point,  would a view like this be more appropriate?

screen shot 2017-11-20 at 11 33 03


_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71737) by @beardhatcode on Mon Nov 20 2017 at 11:33._

**Comment by bmesuere** [comment]
That's better indeed, but I'm still not convinced of splitting the width in 3 and have multi-line table rows for longer names.
Depending on the aspect ratio of the quick go images, we could limit the table to a bootstrap width of 8 or 9, and use the remaining 3 or 4 to display the image next to it.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71738) by @bmesuere on Mon Nov 20 2017 at 11:40._

**Comment by beardhatcode** [comment]
The aspect ratio strongly depended on the terms that are shown. Sometimes it is very wide, sometimes very narrow.

https://app.moqups.com/unipeptMocks/Tk0MFESOKg/view/page/ad64222d5
screen shot 2017-11-20 at 12 07 46


_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71739) by @beardhatcode on Mon Nov 20 2017 at 12:09._

**Comment by beardhatcode** [comment]
**Layout of GO-terms**: Also includes tooltips indicating the count in percent
screen shot 2017-11-20 at 17 49 22

**Clicking on a "QuickGO chart opens it in modal**: clicking on the chart in the modal opens it in a new tab (useful in case it would still be too small but a better solution is needed)
shot


_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71749) by @beardhatcode on Mon Nov 20 2017 at 18:02._

**Comment by beardhatcode** [comment]
Added treeview


_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71771) by @beardhatcode on Tue Nov 21 2017 at 17:23._

**Comment by beardhatcode** [comment]
@bmesuere combining the EC tree and the table gives rise to a UX problem: If you want to scroll to the table, you zoom the EC tree rather than scrolling.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71772) by @beardhatcode on Tue Nov 21 2017 at 17:26._

**Comment by bmesuere** [comment]
Great, this is beginning to take shape!

- [x] something goes wrong for the peptide GGVEPQSETVWR on my local version
```
Uncaught TypeError: Cannot read property 'values' of undefined
    at setUpGO (sequence_show.js:252)
    at setUpFA (sequence_show.js:114)
    at initSequenceShow (sequence_show.js:23)
    at HTMLDocument.<anonymous> (equateIL:415)
    at mightThrow (jquery.js:3583)
    at process (jquery.js:3651)
 ```
- should we display counts or percentages or both in the table?
- [x] maybe switch the background color of the table cells to the "border gray" color instead of pink-ish
- [x] cursor should change when hovering over thumbnails to indicate they can be clicked (maybe even add a title tag)
- [x] add a chevron-down/up before the see more/less text, remove the dotted line and don't show the text if not needed ("show 0 more rows")
- [x] the tooltip is a good idea, but I would swap the name and the number. Maybe play with font-size and color improve the hierarchy.
- for higher GO thumbnails, there is a vertical gap between the tables at the left. Swapping the position of the tables and images might fix this, but I don't know if this looks good.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71803) by @bmesuere on Wed Nov 22 2017 at 19:51._

**Comment by beardhatcode** [comment]
- I still have a TODO on my blackboard to add a message that there are no GO or EC terms. So that will be the cause of that error I assume.

- I think we should only show counts, since many peptides don't have GO numbers associated with them the percentage may give an inaccurate depiction. 
   - Maybe we should add a text like "XX% of matched peptides have GO term annotations"  
   - Maybe this can be added as badge to the tab?


_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71805) by @beardhatcode on Wed Nov 22 2017 at 22:03._

**Comment by bmesuere** [comment]
- [x] oplossing zoeken voor hoogte go figuren                                           
- [x] percentage is fout bij 3 termen die in alle hits voorkomen                        
- [x] bij ec positie figuur en tabel omwisselen en uitklapbaar maken                    
- [x] variabel aantal nodes uitklappen bij ec figuur                                    
- [x] export van de data/figuren                   
   - [x] data                                         
   - [x] EC tree                                  
   - [ ] ~~QuickGo img~~ (download links to other domains do not work)                                      
- [x] bij ec deelstukken benoemen? Evt in tooltips

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71842) by @bmesuere on Fri Nov 24 2017 at 09:51._

**Comment by beardhatcode** [comment]
@bmesuere does [unipept/unipept-visualizations](https://github.ugent.be/unipept/unipept-visualizations) already have a feature to open and close certain parts of the tree?

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71849) by @beardhatcode on Fri Nov 24 2017 at 16:28._

**Comment by bmesuere** [comment]
I think you can call the `click` function with the node as parameter: https://github.com/unipept/unipept-visualizations/blob/master/src/treeview/treeview.js#L315
If that doesn’t work, let me know so I can add such function.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71850) by @bmesuere on Fri Nov 24 2017 at 16:45._

**Comment by beardhatcode** [comment]
Added save button
save btn


_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71893) by @beardhatcode on Mon Nov 27 2017 at 15:24._

**Comment by beardhatcode** [comment]
- [ ] Create a PR for the content type change

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71894) by @beardhatcode on Mon Nov 27 2017 at 15:27._

**Comment by beardhatcode** [comment]
Last commit relies on PR https://github.ugent.be/unipept/unipept-visualizations/pull/21 to be accepted

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71912) by @beardhatcode on Tue Nov 28 2017 at 11:12._

**Comment by bmesuere** [comment]
I will try to merge that early this afternoon. I'll make similar methods available in the 2 other visualisations as well, update the build tools and release a new version.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71914) by @bmesuere on Tue Nov 28 2017 at 11:47._

**Comment by bmesuere** [comment]
@beardhatcode you can update the visualisation lib to 1.6.0 using yarn.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-71919) by @bmesuere on Tue Nov 28 2017 at 15:13._

**Comment by beardhatcode** [comment]
I need a place to put [these lines](https://github.ugent.be/unipept/unipept/blob/feature/functionalAnalysisPept2FA/app/controllers/mpa_controller.rb#L71-L79) to get EC number data. 

I'll probably create a separate controller because this functionality is also needed for mpa

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-72315) by @beardhatcode on Mon Dec 11 2017 at 17:18._

**Comment by beardhatcode** [comment]
EC annotations in the treeview now have tooltips. The data is kept in a global variable (static var of ECNumbers) that cashes the names of all EC numbers.
screenshot_20171213_161026

- [ ] Note to self: check out of order execution of getting the data


_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-72335) by @beardhatcode on Wed Dec 13 2017 at 16:15._

**Comment by beardhatcode** [comment]
- [x] Place `postJSON` in `utils.js`
- [ ] Tooltips should be generalised

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-72365) by @beardhatcode on Fri Dec 15 2017 at 17:03._

**Comment by bmesuere** [comment]
image

- [ ] Because of the vertical spacing, the hierarchy isn't clear. Headers should be closer to their content. Increase the top margin of these headers or decrease the bottom margin.
- [x] The "name" table header floats a bit
- [x] I would always show the "save CSV" (rename to "save as CSV") instead of only on hover
- [x] Link the go terms to quick go instead of amigo

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-72399) by @bmesuere on Mon Dec 18 2017 at 15:03._

**Comment by bmesuere** [comment]
Tooltip layout mockup
image

```
<div id="tooltip" class="tip">
    <h4 class="tooltip-go-title">
        <span class="tooltip-go-title-name">intracellular</span>
        <span class="tooltip-go-title-term small"> GO:0005622</span>
    </h4>
    <span class="tooltip-go-domain">Cellular Component</span>
    <div class="tooltip-go-text">Specificly assigned to 5 (1.2%)
        annotated matches</div>
</div>
```
```
h4.tooltip-go-title {
    margin-top: 6px;
    line-height: initial;
}
title-term.small {
    font-weight: 300;
}
span.tooltip-go-domain {
    font-size: small;
    color: our internal gray variable;
}
.tooltip-go-text {
    margin-top: 8px;
}
```
- [ ] increase tip padding to 8px
- [x] titleize the name? 
  - no
- [x] tweak text "Assigned to xx% (#) of matched proteins"

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-72400) by @bmesuere on Mon Dec 18 2017 at 15:10._

**Comment by beardhatcode** [comment]
image


_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-73209) by @beardhatcode on Tue Feb 13 2018 at 11:42._

**Comment by beardhatcode** [comment]
@bmesuere I think this is ready.

For the tooltip padding, is it ok to chenge the padding value of the `.tip` class?

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-73248) by @beardhatcode on Wed Feb 14 2018 at 11:49._

**Comment by bmesuere** [comment]
Ok, I'll review this before the end of the week.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-73249) by @bmesuere on Wed Feb 14 2018 at 11:57._

**Comment by beardhatcode** [comment]
I'm going to make some more adjustments, to have a better base for MPA.
@bmesuere


I added the comment to the wrong PR. Have you already started looking at it. 

The main change is that I moved some code to the GoTerns class and I removed some magic values. 

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-73291) by @beardhatcode on Thu Feb 15 2018 at 17:08._

File: app/assets/javascripts/components/amounttable.js [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+        this.tooltip = d3.select("body")
+            .append("div")
+            .attr("class", "tip")
+            .style("position", "absolute")
+            .style("z-index", "10")
+            .style("visibility", "hidden")
+            .style("background-color", "white")
+            .style("padding", "2px")
+            .style("border", "1px solid #dddddd")
+            .style("border-radius", "3px;");
```
Comment (bmesuere) [inline app/assets/javascripts/components/amounttable.js]: These styles seem to be fixed, so why not put them in the css file?

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13712) by @bmesuere on Mon Feb 19 2018 at 13:36._

File: app/assets/javascripts/components/amounttable.js [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+    /**
+     * Make the table visualization
+     */
+    draw() {
+        const [thead, tbody] = this.buildTable();
+        this.buildHeader(thead);
+        if (this.data.length != 0) {
```
Comment (bmesuere) [inline app/assets/javascripts/components/amounttable.js]: style: always use `!==` or `===`

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13713) by @bmesuere on Mon Feb 19 2018 at 13:37._

File: app/assets/javascripts/components/amounttable.js [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+            tbody.append("tr").append("td")
+                .attr("colspan", this.header.length)
+                .style("text-align", "center")
+                .text("No data");
+        }
+    }
+
+    /**
+     * Create a table in the target element (without creating it first)
+     * @return {d3.selection}  Table head and table body
```
Comment (bmesuere) [inline app/assets/javascripts/components/amounttable.js]: return type should be `d3.selection[]`

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13714) by @bmesuere on Mon Feb 19 2018 at 13:39._

File: app/assets/javascripts/components/amounttable.js [hunk 4]
```diff
@@ -0,0 +1,313 @@
+import {toCSVString, downloadData} from "../utils.js";
+
+
+/**
+ * @typedef {Object} AmountTableSettings
+ * @property {string}   el        A CSS selector of the target container
```
Comment (bmesuere) [inline app/assets/javascripts/components/amounttable.js]: Is this really a string? Lower in this file, `append` is called on it. Is it a jQuery object? If so, our convention is to use variable names starting with a `$`.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13715) by @bmesuere on Mon Feb 19 2018 at 13:43._

File: app/assets/javascripts/components/amounttable.js [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+
+    /**
+     * Create a table in the target element (without creating it first)
+     * @return {d3.selection}  Table head and table body
+     */
+    buildTable() {
+        if (this.table === null) {
+            this.table = this.el.append("table").attr("class", "table table-condensed table-amounttable");
+        }
+        this.table.html(); // empty the tables HTML
```
Comment (bmesuere) [inline app/assets/javascripts/components/amounttable.js]: Without parameter, this returns the html (in both jquery and d3, I don't know what `this.table` is). An empty string should be added as parameter.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13716) by @bmesuere on Mon Feb 19 2018 at 13:46._

File: app/assets/javascripts/components/amounttable.js [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+     * @param {d3.selection} tbody table body
+     */
+    buildRow(tbody) {
+        let rows = tbody.selectAll("tr")
+            .data(this.collapsed ? this.data.slice(0, this.limit): this.data);
+
+        rows.exit().remove(); // remove rows that are no longer needed
+
+        let row = rows.enter().append("tr");
+        for (let colSpec of this.contents) {
```
Comment (bmesuere) [inline app/assets/javascripts/components/amounttable.js]: When using for ... of loops, use `const` over `let` if the variable isn't reassigned in the loop block. Change on several places in this file.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13717) by @bmesuere on Mon Feb 19 2018 at 13:54._

File: app/assets/javascripts/components/ecnumbers.js [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+     * @param  {[FACounts]} ec list of EC numbers with their counts
+     */
+    constructor({numAnnotatedPeptides = null, data}) {
+        this.numTotalSet = numAnnotatedPeptides;
+        this.data = Array.from(data).sort((a, b) => (b.value - a.value));
+        this.ec = new this[addMissing](data);
+        ECNumbers[addNames](data);
+
+        // Fetch names in the background, not needed yet
+        setTimeout(()=>{
```
Comment (bmesuere) [inline app/assets/javascripts/components/ecnumbers.js]: spaces around `=>`, strange there is no eslint rule for this.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13718) by @bmesuere on Mon Feb 19 2018 at 13:59._

File: app/assets/javascripts/components/ecnumbers.js [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+import {postJSON} from "../utils.js";
+/**
+ * @typedef {Object} FACounts
+ * @property {number} value count
+ * @property {string} name  The name of the GO/EC number
+ * @property {string} code  The code of the GO/EC number
+ */
+
+// const for private methods
+const addMissing = Symbol("[addMissing]");
```
Comment (bmesuere) [inline app/assets/javascripts/components/ecnumbers.js]: why are symbols used for this?

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13719) by @bmesuere on Mon Feb 19 2018 at 14:01._

File: app/assets/javascripts/components/ecnumbers.js [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+     * Create an EC tree
+     *
+     * The numbers are first sorted form general to specific (number of - at the
+     * end). Then the tree is built by adding the count to all ancestors.
+     *
+     * @param {string} target the element to put the tree in (will be cleared)
+     * @param {object} treeViewOptions options to pass to the treeview
+     * @return {TreeView} the created tree
+     */
+    createTree(target, treeViewOptions) {
```
Comment (bmesuere) [inline app/assets/javascripts/components/ecnumbers.js]: This function is quite long, consider splitting it up if possible.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13720) by @bmesuere on Mon Feb 19 2018 at 14:03._

File: app/assets/javascripts/components/goterms.js [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+    /**
+     * @return {int} number of annotated peptides
+     */
+    getTotalSetSize() {
+        return this.numTotalSet;
+    }
+
+    /**
+     * s
```
Comment (bmesuere) [inline app/assets/javascripts/components/goterms.js]: missing text, probably due to pressing ctrl + s

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13721) by @bmesuere on Mon Feb 19 2018 at 14:06._

File: app/assets/javascripts/sequence_show.js [hunk 11]
```diff
@@ -105,9 +131,255 @@ function initSequenceShow(data) {
         }
     }
 
+    function setUpFA(fa) {
+        setUpGO(fa.go);
+        setUpEC(fa.ec);
+    }
+
+    function setUpEC({numAnnotatedPeptides, data}) {
+        let ecResultset = new ECNumbers({numAnnotatedPeptides, data});
```
Comment (bmesuere) [inline app/assets/javascripts/sequence_show.js]: `let` can be `const` on various places in this file.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13723) by @bmesuere on Mon Feb 19 2018 at 14:20._

File: app/assets/javascripts/sequence_show.js [hunk 12]
```diff
@@ -105,9 +131,255 @@ function initSequenceShow(data) {
...
     /**
-     * Inits the lineage tree
+     * Create the EC treeview
+     *
+     * @param {ECNumbers} ecResultSet  A `ECNumbers` summary
+     * @return {TreeView} The created treeview
      */
+    function setUpECTree(ecResultSet) {
+        let tree = ecResultSet.createTree("#ec-treeview", {
+            width: 916,
```
Comment (bmesuere) [inline app/assets/javascripts/sequence_show.js]: magic numbers

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13724) by @bmesuere on Mon Feb 19 2018 at 14:27._

File: app/assets/javascripts/utils.js [hunk 13]
```diff
@@ -31,6 +31,25 @@ function brightness(rgb) {
     return rgb.r * 0.299 + rgb.g * 0.587 + rgb.b * 0.114;
 }
 
+/**
```
Comment (bmesuere) [inline app/assets/javascripts/utils.js]: Include link to https://caniuse.com/#feat=download in comments. Basically IE11 only

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13725) by @bmesuere on Mon Feb 19 2018 at 14:29._

File: app/controllers/sequences_controller.rb [hunk 14]
```diff
@@ -114,6 +114,29 @@ def show
 
     @title = "Tryptic peptide analysis of #{@original_sequence}"
 
+    @pept_with_go = 0
+    @pept_with_ec = 0
+    go_counts = Hash.new 0
+    @go_summary = {}
+    @ec_summary = Hash.new 0
+    @entries.each do |entery|
```
Comment (bmesuere) [inline app/controllers/sequences_controller.rb]: typo: entery

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13727) by @bmesuere on Mon Feb 19 2018 at 14:34._

File: app/controllers/sequences_controller.rb [hunk 15]
```diff
@@ -114,6 +114,29 @@ def show
...
+      go_terms = entery.go_cross_references.map(&:go_term)
```
Comment (bmesuere) [inline app/controllers/sequences_controller.rb]: you can request the go_terms directly from the entries

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13729) by @bmesuere on Mon Feb 19 2018 at 14:36._

File: app/controllers/sequences_controller.rb [hunk 16]
```diff
@@ -114,6 +114,29 @@ def show
 
     @title = "Tryptic peptide analysis of #{@original_sequence}"
 
+    @pept_with_go = 0
```
Comment (bmesuere) [inline app/controllers/sequences_controller.rb]: I think this counting code can be made shorter (and clearer variable names?) or should be moved to a separate function. Maybe http://ruby-doc.org/core-2.5.0/Enumerable.html#method-i-group_by can help.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13730) by @bmesuere on Mon Feb 19 2018 at 14:40._

File: app/assets/javascripts/components/amounttable.js (hunk 4 above)
Comment (beardhatcode) [inline app/assets/javascripts/components/amounttable.js]: String is wrong, it is a `d3.selection` what is the convention for those objects?

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13731) by @beardhatcode on Mon Feb 19 2018 at 15:02._

File: app/assets/javascripts/components/ecnumbers.js (hunk 8 above)
Comment (beardhatcode) [inline app/assets/javascripts/components/ecnumbers.js]: to create "private" static methods, I'll make them simple static methods. In retrospect symbols were not needed

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13732) by @beardhatcode on Mon Feb 19 2018 at 15:20._

**Comment by beardhatcode** [comment]
@bmesuere I handled all remarks

There has been a change in the notation of the percentages in the EC tree. To be more correct, I removed them form EC nodes in the tree who's value is computed by summing the children. Placing a percentage here would be misleading because it would be the percentage of the occurrences of EC terms and not of the number of annotated peptides.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-73485) by @beardhatcode on Thu Feb 22 2018 at 15:37._

File: app/assets/javascripts/sequence_show.js [hunk 17]
```diff
@@ -1,7 +1,29 @@
-import {addCopy, logToGoogle, triggerDownloadModal} from "./utils.js";
+import {addCopy, logToGoogle, triggerDownloadModal, stringTitleize, numberToPercent} from "./utils.js";
 import {showInfoModal} from "./modal.js";
-
```
Comment (bmesuere) [inline app/assets/javascripts/sequence_show.js]: typo: panel

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13896) by @bmesuere on Wed Feb 28 2018 at 15:10._

Comment (beardhatcode) [inline app/assets/javascripts/sequence_show.js]: @bmesuere , All pannels are now panels

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13911) by @beardhatcode on Fri Mar 02 2018 at 09:49._

**Comment by beardhatcode** [comment]
The changes form this PR in #623 .

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#issuecomment-77563) by @beardhatcode on Mon May 28 2018 at 14:21._

**Pull request closed without merging**